auth/service.cc

```diff
@@ -627,4 +633,64 @@ future<std::vector<permission_details>> list_filtered_permissions(
     });
 }
 
+future<> migrate_to_auth_v2(cql3::query_processor& qp, ::service::raft_group0_client& g0, start_operation_func_t start_operation_func, abort_source& as) {
+    // FIXME: if this function fails it may leave partial data in the new tables
+    // that should be cleared
+    auto gen = [&qp] (api::timestamp_type& ts) -> mutations_generator {
+        for (const auto& cf_name : std::vector<sstring>{
+            "roles", "role_members", "role_attributes", "role_permissions"}) {
+            schema_ptr schema;
+            try {
+                schema = qp.db().find_schema(meta::legacy::AUTH_KS, cf_name);
+            } catch (data_dictionary::no_such_column_family&) {
+                continue; // some tables might not have been created if they were not used
+            }
+            auto rows = co_await qp.execute_internal(
+                format("SELECT * FROM {}.{}", meta::legacy::AUTH_KS, cf_name),
+                db::consistency_level::ALL,
+                {},
+                cql3::query_processor::cache_internal::no);
+            if (rows->empty()) {
+                continue;
+            }
+            std::vector<sstring> col_names;
+            for (const auto& col : schema->all_columns()) {
+                col_names.push_back(col.name_as_cql_string());
+            }
+            auto col_names_str = boost::algorithm::join(col_names, ", ");
+            sstring val_binders_str = "?";
+            for (size_t i = 1; i < col_names.size(); ++i) {
+                val_binders_str += ", ?";
+            }
+            for (auto row = rows->begin(); row != rows->end(); ++row) {
+                std::vector<data_value_or_unset> values;
+                for (const auto& col : schema->all_columns()) {
+                    if (row->has(col.name_as_text())) {
+                        values.push_back(
+                            col.type->deserialize(row->get_blob(col.name_as_text())));
+                    } else {
+                        values.push_back(unset_value{});
+                    }
+                }
+                auto muts = co_await qp.get_mutations_internal(
+                    format("INSERT INTO {}.{} ({}) VALUES ({})",
+                        db::system_auth_keyspace::NAME,
+                        cf_name,
+                        col_names_str,
+                        val_binders_str),
+                    internal_distributed_query_state(),
+                    ts,
+                    std::move(values));
+                for (auto& mut : muts) {
```

Comment:
Similarly as in the other use case of `mutations_generator`, we should verify there's a single mutation there.
- It's a single-row insert so it will be a single mutation
- if it could be multiple mutations, then we'd be using potentially wrong timestamps as guard could be retaken during this loop

Reply:
ok